src/python/Utils/TokenManager.py

```diff
@@ -0,0 +1,135 @@
+
+
+def getToken(name=None):
+    """
+    Get IAM token either from environment or file name
+    :param name: ether file name containing token or environment name which hold the token value.
+    If not provided it will be assumed to read token from IAM_TOKEN environment.
+    :return: token or None
+    """
+    if name and os.path.exists(name):
+        token = None
+        with open(name) as istream:
+            token = istream.read()
+        return token
+    if name:
+        return os.environ.get(name)
+    return os.environ.get("IAM_TOKEN")
+
+
+def tokenData(token, url=None, audUrl=None):
+    """
+    inspect and extract token data
+    :param token: token string
+    :param url: IAM provider URL, by default it is https://cms-auth.web.cern.ch/jwk
+    :param audUrl: audience string, by default it is https://wlcg.cern.ch/jwt/v1/any
+    """
+    if not token or not jwt:
+        return {}
+    if isinstance(token, str):
+        token = bytes(token, 'utf-8')
+    if not url:  # use CMS IAM
+        url = "https://cms-auth.web.cern.ch/jwk"
+    jwksClient = jwt.PyJWKClient(url)
+    signingKey = jwksClient.get_signing_key_from_jwt(token)
+    key = signingKey.key
+    headers = jwt.get_unverified_header(token)
+    alg = headers.get('alg', 'RS256')
+    if not audUrl:  # use WLCG
+        audUrl = "https://wlcg.cern.ch/jwt/v1/any"
+    data = jwt.decode(
+        token,
+        key,
+        algorithms=[alg],
+        audience=audUrl,
+        options={"verify_exp": True},
+        )
+    return data
+
+
+def isValidToken(token):
+    """
+    check if given token is valid or not
+
+    :param token: token string
+    :return: true or false
+    """
+    tokenDict = {}
+    try:
+        tokenDict = tokenData(token)
+    except:
+        traceback.print_exc()
+        raise
+    exp = tokenDict.get('exp', 0)
+    if not exp or exp < time.time():
+        return False
+    return True
+
+
+class TokenManager():
+    """
+    TokenManager class handles IAM tokens
+    """
+
+    def __init__(self, name=None, url=None, audUrl=None, logger=None):
+        """
+        Token manager reads IAM tokens either from file or env.
+        It caches token along with expiration timestamp.
+        By default the env variable to use is IAM_TOKEN.
+        :param name: string representing either file or env where we should read token from
+        :param url: IAM provider URL, by default it is https://cms-auth.web.cern.ch/jwk
+        :param audUrl: audience string, by default it is https://wlcg.cern.ch/jwt/v1/any
+        :param logger: logger object or none to use default one
+        """
+        self.name = name
+        self.url = url
+        self.audUrl = audUrl
+        self.expire = 0
+        self.token = None
+        self.logger = logger if logger else logging.getLogger()
+        try:
+            self.token = self.getToken()
+        except Exception as exc:
+            self.logger.exception(str(exc))
```

Comment:
Can you add some more context to this error message? Something like:
```
self.logger.exception("Failed to get token. Details: %s", str(exc))
```
?

Reply:
done